libsoftwarecontainer/src/gateway/networkgateway.cpp

```diff
@@ -41,28 +40,173 @@ NetworkGateway::~NetworkGateway()
 
 ReturnCode NetworkGateway::readConfigElement(const json_t *element)
 {
-    if (!read(element, "internet-access", m_internetAccess)) {
-        log_error() << "either key \"internet-access\" missing or not a bool in json configuration";
+    Entry e;
+    if (!read(element, "type", e.type)) {
+        log_error() << "No type specified in network config.";
         return ReturnCode::FAILURE;
     }
 
-    if (m_internetAccess) {
-        log_info() << "Internet access will be enabled";
+    if (e.type != "INCOMING" && e.type != "OUTGOING") {
+        log_error() << e.type << " is not a valid type ('INCOMING' or 'OUTGOING')";
+        return ReturnCode::FAILURE;
+    }
+
+    const json_t *prio = json_object_get(element, "priority");
+    if (prio == nullptr) {
+        log_error() << "No priority specified in network config.";
+        return ReturnCode::FAILURE;
+    }
 
-        if (!read(element, "gateway", m_gateway)) {
-            log_error() << "either key \"gateway\" missing or not a string in json configuration";
+    e.priority = json_integer_value(prio);
+    if (e.priority < 1) {
+        log_error() << "Priority can not be less than 1";
+        return ReturnCode::FAILURE;
+    }
+
+
+    const json_t *rules = json_object_get(element, "rules");
+
+    if (rules == nullptr) {
+        log_error() << "No rules specified";
+        return ReturnCode::FAILURE;
+    }
+
+    if (!json_is_array(rules)) {
+        log_error() << "Rules not specified as an array";
+        return ReturnCode::FAILURE;
+    }
+
+    size_t ix;
+    json_t *val;
+    json_array_foreach(rules, ix, val) {
+        if (json_is_object(val)) {
+            if (isError(parseRule(val, e.rules))) {
+                log_error() << "Could not parse rule config";
+                return ReturnCode::FAILURE;
+            }
+        } else {
+            log_error() << "formatting of rules array is incorrect.";
+            return ReturnCode::FAILURE;
         }
+    }
+
+    std::string readTarget;
+    if (!read(element, "default", readTarget)) {
+        log_error() << "No default target specified or default target is not a string.";
+        return ReturnCode::FAILURE;
+    }
 
-        if (m_gateway.length() == 0) {
-            log_error() << "Value for \"gateway\" key is an empty string";
+    e.defaultTarget = parseTarget(readTarget);
+    if (e.defaultTarget == Target::INVALID_TARGET) {
+        log_error() << "Default target '" << readTarget << "' is not a supported target.Invalid target.";
+        return ReturnCode::FAILURE;
+    }
+
+    m_entries.push_back(e);
+
+    // --- TEMPORARY WORKAROUND ---
+    // in the wait of activate() being rewritten
+    if (e.defaultTarget == Target::ACCEPT) {
+        m_internetAccess = true;
+        m_gateway = "10.0.3.1";
+    }
+    // ----------------------------
+    return ReturnCode::SUCCESS;
+}
+
+ReturnCode NetworkGateway::parseRule(const json_t *element, std::vector<Rule> &rules)
```

Comment:
I've been thinking about that. It is untrivial as it is so closely coupled with parsing the rule. On the other hand I agree with the function being to big.